Data Linker Update @ Glue Solara - Create Multiple Link types

ChesleyC opened the pull request:
Data Linker is a component in Glue Solara web UI, which implements Qt's LinkEditor functionality, by importing Glue-core's components.
- Data Linker components are now allowed to create multiple link types, including advanced links from glue-core and glue-qt.
- Allow editing the attributes of each links.

Main structures:
  1. Registry caching system (module-level, runs once)
  2. AdvancedLinkMenu component 
  3. Helper functions (legacy)
  4. stringify_links() display function
  5. Linker main orchestrator
  6. LinkSelector component
  7. CurrentLinksSelector component
  8. QtStyleLinkDetailsPanel (THE CORE)
  9. _get_selected_link_info() helper

<img width="1563" height="1567" alt="image" src="https://github.com/user-attachments/assets/98b2ba34-8747-4bee-8e61-4cd51b0ef409" />




Advanced Links:


<img width="1361" height="367" alt="image" src="https://github.com/user-attachments/assets/0f313181-65c2-416b-b3cd-54beb386a46f" />

<img width="1321" height="519" alt="image" src="https://github.com/user-attachments/assets/f1179274-f127-4356-8e18-7fbea0db8e28" />

<img width="1371" height="377" alt="image" src="https://github.com/user-attachments/assets/ebbd3a4a-dc02-48a4-a8a4-6ceecb1f2dfd" />

ChesleyC committed:
Update linker.py

ChesleyC committed:
Advanced Link Done - need verify

ChesleyC committed:
Advanced Link Done - 1st verified

ChesleyC committed:
Advanced Link UI Width Changed

ChesleyC committed:
fix: link editing bugs (LinkSame, dedup, multi-param)

ChesleyC committed:
fix: link editing bugs (LinkSame, dedup, multi-param) - Join Link ex

ChesleyC committed:
2025.10.07-starting on Link GUI

ChesleyC committed:
nonGUI: fix: 'join', '3d galactocentric', link editing / removal

ChesleyC committed:
Linker (without GUI) - tidy version

ChesleyC committed:
add accidentally deleted file

ChesleyC committed:
remove testing dataset

ChesleyC committed:
restore the an original file

ChesleyC committed:
fix: remove last F841 (original_position) and apply format

Carifio24 commented on glue_solara/linker.py:
```diff
@@ -1,60 +1,755 @@
+        if hasattr(helper_class, "cid_independent") and helper_class.cid_independent:
+            links = helper_class(data1=data1, data2=data2)
+        else:
+            from glue_qt.dialogs.link_editor.state import EditableLinkFunctionState
```
One thing that we definitely don't want to do is import from `glue-qt`, as we don't want to have that as a dependency in a Solara context. We'll either need to recreate the functionality of that state class here, or think about moving it upstream to `glue-core`, which is a dependency here.

Carifio24 commented on glue_solara/linker.py:
```diff
@@ -1,60 +1,755 @@
+    Returns:
+        str: Human-readable link description
+    """
+    try:
```
I appreciate the effort in this function, but I think a lot of this can be done more simply. For example, all of the coordinate function link types already have a display attribute that we can use - see e.g. [here](https://github.com/glue-viz/glue/blob/main/glue/plugins/coordinate_helpers/link_helpers.py#L42). It would be better to use that than to repeat all of those constants here.

Carifio24 commented on glue_solara/linker.py:
```diff
@@ -1,60 +1,755 @@
+            cid2_label = getattr(link._cid2, "label", str(link._cid2))
+            return f"{cid1_label} <-> {cid2_label}"
+
+        elif type(link).__name__ == "JoinLink":
```
Instead of checking the `__name__`, we should just directly check the type - i.e. import `JoinLink` and then check `isinstance(link, JoinLink)`

Carifio24 commented on glue_solara/linker.py:
```diff
@@ -1,60 +1,755 @@
+        elif type(link).__name__ == "JoinLink":
+            return str(link)
+
+        elif "coordinate_helpers" in str(type(link)):
```
This block is concerned with the astronomy links, so I think you could instead check whether `isinstance(link, BaseCelestialMultiLink)`. But this block might not be needed - see my comment above about simplifying this function.

Carifio24 commented on glue_solara/linker.py:
```diff
@@ -1,60 +1,755 @@
             data_collection[selected_data2.value],
             data_collection[selected_data2.value].components[selected_row2.value],
         )
+        time.sleep(0.1)
```
Is there a reason for the `sleep` call here? Is there a glue operation that we need to finish?

Carifio24 commented on glue_solara/linker.py:
```diff
@@ -63,24 +758,1327 @@ def LinkSelector(
+                else:
+                    return
+
+                original_link_type = type(link).__name__
```
Similar to comments above - rather than converting type names to strings and trying to parse them, we should just check types directly.

Carifio24 commented on glue_solara/linker.py:
```diff
@@ -63,24 +758,1327 @@ def LinkSelector(
+
+
+@solara.component
+def QtStyleLinkDetailsPanel(
```
I understand why you named it like this, but I don't think we want the class name specifically referencing Qt. This can just be `LinkDetailsPanel`, or something similar

Carifio24 commented on glue_solara/linker.py:
```diff
@@ -63,24 +758,1327 @@ def LinkSelector(
+                to_data = link.data2
+
+                if dataset == 1:
+                    if new_attr_index < len(from_data.components):
```
A lot of these two blocks are redundant. It seems like you could define `data = from_data if dataset == 1 else to_data`, get `new_component` from that, and then just write all of the logic once.

Carifio24 commented on glue_solara/linker.py:
```diff
@@ -63,24 +758,1327 @@ def LinkSelector(
+    if len(data_collection) == 0:
+        return solara.Text("No data available")
+
+    def _remove_link():
```
I'm not sure where the link is actually removed in this function (and in my local testing it didn't seem to work). I think we need a call to the link manager's `remove_link` method somewhere in here.

Carifio24 commented on glue_solara/linker.py:
```diff
@@ -1,60 +1,755 @@
+        is_join_request = helper_class_name == "JoinLink"
+
+        try:
+            from glue.dialogs.link_editor.state import LinkEditorState
```
I don't think any of the things that we're importing here are going to be costly to load, so I think we can move all of the imports up to the top of the file.

Carifio24 reviewed (changes requested):
Overall I think this is a good start. I noticed a couple of issues when testing this out locally:
* The "Remove Link" button wasn't working for me - pressing it would de-select the currently selected link, but it didn't remove it from the list. The link was still present after closing and re-opening the linker, so I don't think it's just a case of stale state. I also couldn't find where the link was actually deleted in the callback - see my comment below.
* The "Astronomy" category for the advanced links wasn't present for me. I know it was when we looked at this on Zoom earlier today, so perhaps there's something that you have in a local config that isn't generally happening?

I also have a few overall code suggestions. There's a lot of inspection of the links and their classes here, which I think some degree of is unavoidable. However, there's a lot of things that I think we can do in a more straightforward way - for example, rather than checking if some substring is in the typename of an object, we can just import the type and check if the object is an instance of it. Similar for if we want to check whether something is a subtype of something else - we can just use `issubclass`. This should be less brittle and more readable. I have some more specific comments on this throughout.

ChesleyC commented:
@Carifio24 Hi Jon! I've received the first review. There are many valuable details. I've read all the comments and I'll think about them. Thanks for the thoughtful review on my first push!

I also notice that there is "Requested changes with read-only permissions". If I could help with any permission, such as my fork repo, please let me know!

ChesleyC committed:
Fix: 12 issues at codereview

ChesleyC committed:
fix: 12 issue at code review

Reorder import statements and add missing import for time. Adjust comments for clarity.

ChesleyC committed:
fix: precommit

ChesleyC committed:
fix: adding the missing WCS link back

ChesleyC commented:
Thank you for the thorough review, Jon @Carifio24 ! I've addressed all 12 issues you identified. 

Critical fixes: 
(1) restored the astronomy plugin import that clangtidy removed (line 174), 
(2) added temp_state.links.pop(target_index) to fix the remove link bug (line 935), 
(3) replaced string-based type checking with isinstance() throughout, 
(4) moved all imports to top of file, 
(5) removed the glue-qt dependency. 

While testing, I also discovered and fixed a bug where 3-to-3 coordinate links GalactocentricToGalactic) were displaying as 1-to-1—changed isinstance checks from BaseCelestialMultiLink to BaseMultiLink in 5 locations to properly catch all multi-dimensional transforms. 

Additionally added wcs_autolinking import to restore WCS Link visibility in the advanced menu. 

All changes verified against glue-core source code and CI tests passing. Please review my latest commit!
